mindmeld/components/entity_resolver.py

```diff
@@ -562,38 +776,829 @@ def _predict_exact_match(self, entity):
+                Can be set to token_embeddings to get wordpiece token embeddings.
+            convert_to_numpy (bool): If true, the output is a list of numpy vectors. Else, it is a
+                list of pytorch tensors.
+            convert_to_tensor (bool): If true, you get one large tensor as return. Overwrites any
+                setting from convert_to_numpy
+            device: Which torch.device to use for the computation
+            concat_last_n_layers (int): number of hidden outputs to concat starting from last layer
+            normalize_token_embs (bool): if the (sub-)token embs are to be individually normalized
+
+        Returns:
+            (Union[List[Tensor], ndarray, Tensor]): By default, a list of tensors is returned.
+                If convert_to_tensor, a stacked tensor is returned. If convert_to_numpy, a numpy
+                matrix is returned.
+        """
+
+        if concat_last_n_layers != 1:
+            assert 1 <= concat_last_n_layers <= self._num_layers(self.transformer_model.auto_model)
+
+        self.transformer_model.eval()
+        if show_progress_bar is None:
+            show_progress_bar = (
+                logger.getEffectiveLevel() == logging.INFO or
+                logger.getEffectiveLevel() == logging.DEBUG
+            )
+
+        if convert_to_tensor:
+            convert_to_numpy = False
+
+        input_was_string = False
+        if isinstance(sentences, str):  # Cast an individual sentence to a list with length 1
+            sentences = [sentences]
+            input_was_string = True
+
+        device = device or self.device
+
+        self.transformer_model.to(device)
+        self.pooling_model.to(device)
+
+        all_embeddings = []
+        length_sorted_idx = np.argsort([self._text_length(sen) for sen in sentences])
+        sentences_sorted = [sentences[idx] for idx in length_sorted_idx]
+
+        for start_index in trange(0, len(sentences), batch_size, desc="Batches",
+                                  disable=not show_progress_bar):
+            sentences_batch = sentences_sorted[start_index:start_index + batch_size]
+            features = self.transformer_model.tokenize(sentences_batch)
+            features = self._batch_to_device(features, device)
+
+            with _get_module_or_attr("torch", "no_grad")():
+                out_features_transformer = self.transformer_model.forward(features)
+                token_embeddings = out_features_transformer["token_embeddings"]
+                if concat_last_n_layers > 1:
+                    _all_layer_embeddings = out_features_transformer["all_layer_embeddings"]
+                    token_embeddings = _get_module_or_attr("torch", "cat")(
+                        _all_layer_embeddings[-concat_last_n_layers:],
+                        dim=-1)
+                if normalize_token_embs:
+                    _norm_token_embeddings = _get_module_or_attr("torch.linalg", "norm")(
+                        token_embeddings, dim=2,
```

Comment:
why dim=2?

Reply:
[batch_size, max_len, emb_dim] are the three (0-base) dimensions of token_embeddings, irrespective of passing in single or multiple inputs. Hence dim=2 is for emb_dim.